### ISSUE TITLE
osquery watchdog does not log to tls logger

### ISSUE THREAD

**Issue by dachin11** [osquery/osquery#6790]
Hi,

It does not appear that the below logging for the osquery client writes to the TLS logger plugin.  We are only using the TLS plugin and when I check the osquery_status.log on fleetdm, there is no mention of RAM/CPU limits being hit when queries are denylisted, even though we have some being denied.  We would really like to see the reason each query is denied in the fleet status logs as we forward this data to Splunk.

Can the client be updated so that these are also logged to the TLS logger?

https://github.com/osquery/osquery/blob/b04da26b9e2819e270ed284f9667c3edb719bf1e/osquery/core/watcher.cpp#L527-L537

**Comment by theopolis** [comment]
Correct, the "watchdog process" is intentionally as small as possible, meaning it has no logger functionality.

The way to implement this is to include a small IPC between the watchdog and osquery. I have wanted to implement this before but there is not quick and easy IPC that spans the platforms we support. I think this is possible but it's a small or medium sized project someone has to work on (looking for volunteers!).

**Comment by dachin11** [comment]
Thanks @theopolis.  I had one of our developers look through the code and we realized that the watchdog does do a syslog() call and then writes to the osquery.WARNING log file.  We also found:

```
 // Since the watchdog cannot use the logger plugin the error message
 // should be logged to stderr and to the system log.

```
The fact that we at least have the messages logged locally via that syslog() call helps troubleshooting for us, at least for Mac/Linux, and we could also find other ways to get the data into Splunk if we wanted to have a separate path to Splunk, though we don't really want to do that.  I'll ask internally if anyone wants to (and has time to) contribute and make an enhancement.

One last question - the local logging gives us data, but since we only use the tls plugin, we don't get anything logged locally for windows.  We don't particularly want to enable the logger plugin as we want to limit where query results are stored.  My question is - if we enable the windows_event_log logger plugin, will that allow the syslog() call to log the reason for the watchdog trip to the Windows event logs?

**Comment by dachin11** [comment]
@theopolis - can you confirm that if we enable the windoes_event_log logger plugin that the log entries that show why the watchdog was tripped (i.e. CPU or RAM) would log to the windows event log?  If not, it sounds like we'll have to enable the filesystem logger.

**Comment by dachin11** [comment]
I got confirmation in slack that enabling the windows_event_log logger plugin won't allow the client to log the watchdog errors.  :(

**Comment by faleon** [comment]
Bump, as this issue has come up for our org as well

**Comment by theopolis** [comment]
The fix for @dachin11's note is to implement logging to Windows Event Log here: https://github.com/osquery/osquery/blob/65fbc8413b27588f2bdaf226aca42f64a8cc6594/osquery/logger/logger.cpp#L543

That means the watchdog still does not log to TLS or any other logger plugins, but will log to the system log.

**Comment by dachin11** [comment]
Thanks @theopolis.  Can you confirm if the fix to enable the windows event logging has been completed in any versions of osquery and that it won't require the filesystem logger to also be enabled?  We want to stick with just the TLS logger.

**Comment by theopolis** [comment]
Nothing has been fixed yet, and I don't know of anyone actively working on this. Looking for help to write the code!

**Comment by dabresua** [comment]
We are also affected. I will start investigating it and, hopefully, come up with a solution.

**Comment by dabresua** [comment]
First of all, this is the first time I colaborate in this project, so I tried to do this blueprint the best way I could. Also, I don't know osquery in deep yet.

I think we could use [message_queue](https://www.boost.org/doc/libs/1_79_0/doc/html/boost/interprocess/message_queue_t.html) from the boost library. It creates a shared memory area to exchange messages between a sender and a listener on different processes.

I've tested it on Linux and Windows, works fine. However, it creates temporary files on Windows and other OS that doesn't support POSIX shared memory objects. [More details](https://www.boost.org/doc/libs/1_79_0/doc/html/interprocess/sharedmemorybetweenprocesses.html#interprocess.sharedmemorybetweenprocesses.sharedmemory.emulation).

My idea is to send the messages from the Watcher process, receive them in osquery and use the standard LOG API. This way, the IPC logs will be created as the rest of the logs.

On the other hand, message_queue will reserve a static memory area for a number of messages with a fixed length. I don't think we should care much about the number (1 or 2 will probably suffice), but the length matters. Support variable length and resizing the area is possible, but requires a lot of work and the code will be more difficult to maintain.

```mermaid
sequenceDiagram
    participant Watcher
    participant message_queue
    participant IPC_logger
    participant Osquery

    Watcher->>+Watcher: Initializes
    Watcher->>+message_queue: Init
    message_queue->>+message_queue: Creates fixed length shared memory
    Osquery->>+Osquery: Initializes
    Osquery->>+IPC_logger: Launches thread
    IPC_logger->>+IPC_logger: Opens fixed length shared memory
    loop
        IPC_logger->>+IPC_logger: Listens
        Watcher->>+Watcher: Serializes message
        Watcher->>+message_queue: send(msg, sizeof(msg), severity);
        alt Free space available
            message_queue->>+message_queue: Move to shared memory
        else No free space available
            message_queue->>+message_queue: Keep it on local until free space
        end
        message_queue->>+IPC_logger: Notify
        IPC_logger->>+IPC_logger: Read most prioritary message
        IPC_logger->>+Osquery: LOG(severity) << msg
    end
```
